File: task4/generic_collection/src/main/java/collection/impl/MyArrayList.java
package collection.impl;

import collection.MyList;
import collection.ToNumberFunction;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Optional;

public class MyArrayList<T> implements MyList<T> {
    private static final int DEFAULT_CAPACITY = 10;
    private int size = 0;
    private Object elements[];

    public MyArrayList() {
        elements = new Object[DEFAULT_CAPACITY];
    }

    public MyArrayList(int capacity) {
        elements = new Object[capacity];
    }

    public MyArrayList(T[] array){
        if (array == null){
            size = 0;
            elements = new Object[DEFAULT_CAPACITY];
        } else {
            size = array.length;
            elements = Arrays.copyOf(array, this.size);
        }

    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean add(T element) {
        if(size == elements.length){
            elements = grow(size + 1);
        }
        elements[size++] = element;
        size = size;
        elements = elements;

        return true;
    }

    @Override
    public boolean add(int index, T t) {
        if(index < 0) return false;

        if(size + 1 >= elements.length){
            elements = grow(size + 1);
        }

        System.arraycopy(elements, index, elements, index + 1, size - index);
        elements[index] = t;
        size++;

        return true;
    }

    private Object[] grow(int minCapacity){
        return Arrays.copyOf(elements, newCapacity(minCapacity));
    }

    private int newCapacity(int oldCapacity){
        return (oldCapacity * 3) / 2 + 1;
    }

    @Override
    public Optional<T> get(int index) {
        if (index >= 0 && index < size) {
            return Optional.ofNullable((T) elements[index]);
        }
        return Optional.empty();
    }

    @Override
    public Optional<T> remove(int index) {
        if (index >= 0 && index < size) {
            Object obj = get(index);
            shiftToLeft(index);
            return Optional.ofNullable((T) obj);
        }
        return Optional.empty();
    }

    @Override
    public boolean remove(T t) {
        if (size == 0) return false;

        boolean removingSuccessful = false;
        for (int i = 0; i < size; i++){
            if ((elements[i] != null && elements[i].equals(t) || (elements[i] == null && t == null))){
                shiftToLeft(i);
                removingSuccessful = true;
            }
        }

        return removingSuccessful;
    }

    private void shiftToLeft(int indexStart) {
        size--;
        if (size <= 0) {
            return;
        }

        if (size != indexStart) {
            System.arraycopy(elements, indexStart + 1, elements, indexStart, size - indexStart);
        }

        elements[size] = null;
    }

    @Override
    public Optional<T> min(Comparator<T> comparator) {
        if (size == 0) return Optional.empty();

        int indexMin = 0;
        for(int i = 0; i < size - 1; i++){
            if(comparator.compare((T) elements[i], (T) elements[i + 1]) > 0){
                indexMin = i;
            }
        }
        return Optional.ofNullable((T) elements[indexMin]);
    }

    @Override
    public Optional<T> max(Comparator<T> comparator) {
        if (size == 0) return Optional.empty();

        int indexMax = 0;
        for(int i = 0; i < size - 1; i++){
            if(comparator.compare((T) elements[i + 1], (T) elements[i]) > 0){
                indexMax = i + 1;
            }
        }
        return Optional.ofNullable((T) elements[indexMax]);
    }

    @Override
    public boolean contains(Object object) {
        return indexOf(object) >= 0;
    }

    @Override
    public int indexOf(Object object) {
        return indexOfRange(object, 0, size);
    }

    private int indexOfRange(Object object, int start, int end) {
        Object[] es = elements;
        if (object == null) {
            for (int i = start; i < end; i++) {
                if (es[i] == null) {
                    return i;
                }
            }
        } else {
            for (int i = start; i < end; i++) {
                if (object.equals(es[i])) {
                    return i;
                }
            }
        }
        return -1;
    }

    @Override
    public void sortBubble(Comparator<T> comparator) {
        for (int i = 0; i < size - 1; i++) {
            for(int j = 0; j < size - i - 1; j++) {
                if(comparator.compare((T) elements[j], (T) elements[j + 1]) > 0) {
                    Object swap = elements[j];
                    elements[j] = elements[j + 1];
                    elements[j + 1] = swap;
                }
            }
        }
    }

    @Override
    public void sortInsertion(Comparator<T> comparator) {
        for (int i = 1; i < size; i++){
            Object value = elements[i];
            int j = i;
            while (j > 0 && comparator.compare((T) elements[j - 1], (T) value) > 0){
                elements[j] = elements[j - 1];
                j--;
            }
            elements[j] = value;
        }
    }

    @Override
    public void sortSelection(Comparator<T> comparator) {
        for (int i = 0; i < size; i++) {
            int pos = i;
            Object min = elements[i];
            for (int j = i + 1; j < size; j++) {
                if (comparator.compare((T) elements[j], (T) min) < 0) {
                    pos = j;
                    min = elements[j];
                }
            }
            elements[pos] = elements[i];
            elements[i] = min;
        }
    }

    @Override
    public Optional<T> set(int index, T element) {
        if ((index < size) && (index >= 0)) {
            Object oldElement = elements[index];
            elements[index] = element;
            return Optional.ofNullable((T) oldElement);
        }
        return Optional.empty();
    }

    @Override
    public Optional<T> sum(ToNumberFunction<? super T> mapper) {
        if (size == 0) return Optional.empty();

        if(mapper.applyAsNumber((T) elements[0]) instanceof Byte){
            long sum = 0;
            for (int i = 0; i < size; i++) sum += (Byte) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if(mapper.applyAsNumber((T) elements[0]) instanceof Short){
            long sum = 0;
            for (int i = 0; i < size; i++) sum += (Short) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Integer){
            long sum = 0;
            for (int i = 0; i < size; i++) sum += (Integer) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Long){
            long sum = 0;
            for (int i = 0; i < size; i++) sum += (Long) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Float){
            double sum = 0;
            for (int i = 0; i < size; i++) sum += (Float) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Double){
            double sum = 0;
            for (int i = 0; i < size; i++) sum += (Double) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        return Optional.empty();
    }

    @Override
    public Optional<T> product(ToNumberFunction<? super T> mapper) {
        if (size == 0) return Optional.empty();

        if(mapper.applyAsNumber((T) elements[0]) instanceof Byte){
            long sum = 1;
            for (int i = 0; i < size; i++) sum *= (Byte) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if(mapper.applyAsNumber((T) elements[0]) instanceof Short){
            long sum = 1;
            for (int i = 0; i < size; i++) sum *= (Short) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Integer){
            long sum = 1;
            for (int i = 0; i < size; i++) sum *= (Integer) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Long){
            long sum = 1;
            for (int i = 0; i < size; i++) sum *= (Long) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Float){
            double sum = 1;
            for (int i = 0; i < size; i++) sum *= (Float) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        else if (mapper.applyAsNumber((T) elements[0]) instanceof Double){
            double sum = 1;
            for (int i = 0; i < size; i++) sum *= (Double) (mapper.applyAsNumber((T) elements[i]));
            return (Optional<T>) Optional.ofNullable(sum);
        }

        return Optional.empty();
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        str.append("[" + (size > 0 ? elements[0].toString() : ""));
        for(int i = 1; i < size; i++){
            str.append(", " + (elements[i] != null ? elements[i].toString() : "null"));
        }
        str.append("]");
        return str.toString();
    }
}

File: task7/MyArrayList/src/main/kotlin/collection/impl/MyArrayList.kt
package collection.impl

import collection.MyList
import collection.ToNumberFunction

class MyArrayList <T> :MyList<T> {

    private var size = 0
    private var elements: Array<Any?> = arrayOfNulls(DEFAULT_CAPACITY)

    constructor()

    constructor(capacity: Int) {
        elements = arrayOfNulls(capacity)
    }

    override fun size() :Int {
        return size
    }

    override fun add(obj: T): Boolean {
        if (size == elements.size) {
            elements = grow(size + 1)!!
        }
        elements[size++] = obj
        size = size
        elements = elements

        return true
    }

    private fun grow(minCapacity: Int): Array<Any?>? {
        return elements.copyOf(newCapacity(minCapacity))
    }

    private fun newCapacity(oldCapacity: Int): Int {
        return oldCapacity * 3 / 2 + 1
    }

    override fun add(index: Int, obj: T): Boolean {
        if (index < 0) return false

        if (size + 1 >= elements.size) {
            elements = grow(size + 1)!!
        }

        System.arraycopy(elements, index, elements, index + 1, size - index)
        elements[index] = obj
        size++

        return true
    }

    override fun get(index: Int): T? {
        if (index >= 0 && index < size) {
            return elements[index] as T
        }

        return null
    }

    override fun remove(index: Int): T? {
        if (index >= 0 && index < size) {
            val obj: Any? = get(index)
            shiftToLeft(index)
            return obj as T?
        }

        return null
    }

    override fun remove(obj: T): Boolean {
        if (size == 0) return false

        var removingSuccessful = false
        for (i in 0 until size) {
            if (elements[i] != null && elements[i] == obj || elements[i] == null && obj == null) {
                shiftToLeft(i)
                removingSuccessful = true
            }
        }

        return removingSuccessful
    }

    private fun shiftToLeft(indexStart: Int) {
        size--
        if (size <= 0) {
            return
        }
        if (size != indexStart) {
            System.arraycopy(elements, indexStart + 1, elements, indexStart, size - indexStart)
        }
        elements[size] = null
    }

    override fun min(comparator: Comparator<in T>): T? {
        if (size == 0) return null

        var indexMin = 0
        for (i in 0 until size - 1) {
            if (comparator.compare(elements[i] as T, elements[i + 1] as T) > 0) {
                indexMin = i
            }
        }
        return elements[indexMin] as T
    }

    override fun max(comparator: Comparator<in T>): T? {
        if (size == 0) return null

        var indexMax = 0
        for (i in 0 until size - 1) {
            if (comparator.compare(elements[i + 1] as T, elements[i] as T) > 0) {
                indexMax = i + 1
            }
        }
        return elements[indexMax] as T
    }

    override fun indexOf(obj: T): Int {
        return indexOfRange(obj, 0, size)
    }

    private fun indexOfRange(obj: T?, start: Int, end: Int): Int {
        val es = elements
        if (obj == null) {
            for (i in start until end) {
                if (es[i] == null) {
                    return i
                }
            }
        } else {
            for (i in start until end) {
                if (obj == es[i]) {
                    return i
                }
            }
        }
        return -1
    }

    override fun contains(obj: T): Boolean {
        return indexOf(obj) >= 0
    }

    override fun set(index: Int, obj: T): T? {
        if (index < size && index >= 0) {
            val oldElement = elements[index]!!
            elements[index] = obj
            return oldElement as T
        }
        return null
    }

    override fun sortBubble(comparator: Comparator<in T>) {
        for (i in 0 until size - 1) {
            for (j in 0 until size - i - 1) {
                if (comparator.compare(elements[j] as T, elements[j + 1] as T) > 0) {
                    val swap = elements[j]!!
                    elements[j] = elements[j + 1]
                    elements[j + 1] = swap
                }
            }
        }
    }

    override fun sortInsertion(comparator: Comparator<in T>) {
        for (i in 1 until size) {
            val value = elements[i]!!
            var j = i
            while (j > 0 && comparator.compare(elements[j - 1] as T, value as T) > 0) {
                elements[j] = elements[j - 1]
                j--
            }
            elements[j] = value
        }
    }

    override fun sortSelection(comparator: Comparator<in T>) {
        for (i in 0 until size) {
            var pos = i
            var min = elements[i]!!
            for (j in i + 1 until size) {
                if (comparator.compare(elements[j] as T, min as T) < 0) {
                    pos = j
                    min = elements[j]!!
                }
            }
            elements[pos] = elements[i]
            elements[i] = min
        }
    }

    override fun sum(mapper: ToNumberFunction<in T>?): T? {
        if (size == 0) return null

        if (mapper!!.applyAsNumber(elements[0] as T) is Byte) {
            var sum: Long = 0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Byte?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Short) {
            var sum: Long = 0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Short?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Int) {
            var sum: Long = 0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Int?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Long) {
            var sum: Long = 0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Long?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Float) {
            var sum: Double = 0.0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Float?)!!.toDouble()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Double) {
            var sum: Double = 0.0
            for (i in 0 until size) sum += (mapper!!.applyAsNumber(elements[i] as T) as Double?)!!.toDouble()
            return sum as T
        }

        return null
    }

    override fun product(mapper: ToNumberFunction<in T>?): T? {
        if (size == 0) return null

        if (mapper!!.applyAsNumber(elements[0] as T) is Byte) {
            var sum: Long = 0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Byte?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Short) {
            var sum: Long = 0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Short?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Int) {
            var sum: Long = 0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Int?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Long) {
            var sum: Long = 0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Long?)!!.toLong()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Float) {
            var sum: Double = 0.0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Float?)!!.toDouble()
            return sum as T
        }

        if (mapper!!.applyAsNumber(elements[0] as T) is Double) {
            var sum: Double = 0.0
            for (i in 0 until size) sum *= (mapper!!.applyAsNumber(elements[i] as T) as Double?)!!.toDouble()
            return sum as T
        }

        return null
    }

    override fun toString(): String {
        val str = StringBuilder()
        str.append("[" + if (size > 0) elements[0].toString() else "")
        for (i in 1 until size) {
            str.append(", " + if (elements[i] != null) elements[i].toString() else "null")
        }
        str.append("]")
        return str.toString()
    }

    companion object {
        private const val DEFAULT_CAPACITY = 10
    }


}